x-pack/plugins/global_search/server/services/search_service.ts

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Observable, timer, merge, throwError } from 'rxjs';
+import { map, takeUntil } from 'rxjs/operators';
+import { KibanaRequest, CoreStart, IBasePath } from 'src/core/server';
+import { GlobalSearchProviderResult } from '../../common/types';
+import { takeInArray } from '../../common/operators';
+import { ILicenseChecker } from '../../common/license_checker';
+
+import { processProviderResult } from '../../common/process_result';
+import { GlobalSearchConfigType } from '../config';
+import { getContextFactory, GlobalSearchContextFactory } from './context';
+import {
+  GlobalSearchResultProvider,
+  GlobalSearchBatchedResults,
+  GlobalSearchFindOptions,
+} from '../types';
+
+/** @public */
+export interface SearchServiceSetup {
+  registerResultProvider(provider: GlobalSearchResultProvider): void;
+}
+
+/** @public */
+export interface SearchServiceStart {
+  find(
+    term: string,
+    options: GlobalSearchFindOptions,
+    request: KibanaRequest
+  ): Observable<GlobalSearchBatchedResults>;
+}
+
+interface SetupDeps {
+  basePath: IBasePath;
+  licenseChecker: ILicenseChecker;
+  config: GlobalSearchConfigType;
+  maxProviderResults?: number;
+}
+
+const defaultMaxProviderResults = 20;
+
+/** @internal */
+export class SearchService {
+  private readonly providers: GlobalSearchResultProvider[] = [];
```

Comment:
I don't necessarily have any problem with it, but we've tended to prefer `Map` from items like this. Any reason not to use a Map? Probably doesn't matter too much since it is an implementation detail.

Reply:
Will switch to `Map` for consistency.